### PR TITLE
Lock validation and persist through graph visitor

### PR DESCRIPTION
**Pull request by ljacomet**
This introduces validation of graph resolution with the lock file.
Added or missing dependencies in the graph compared to the lock
file cause a failure.
The visitor doing the validation is also used for triggering
the persistence of the result.

### CODE REVIEW

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy [hunk 1]
```diff
@@ -341,8 +337,8 @@ dependencies {
         fails 'dependencies'
 
         then:
-        failure.assertHasCause("Dependency lock out of date (strict mode):\n" +
-            "Module missing from lock file: org:bar:1.0")
+        failure.assertHasCause("Dependency lock for configuration 'lockedConf' is out of date:\n" +
+            "\tResolution resolved 'org:bar:1.0' which was not in the lockfile")
```
Comment (melix) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy]: :o: "Resolution resolved" reads strange. Suggestion:

> Resolved 'org:bar:1.0' which wasn't found in the lock file

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy [hunk 2]
```diff
@@ -192,7 +188,8 @@ dependencies {
         fails 'dependencies'
 
         then:
-        failure.assertThatCause(containsText("Lock file contained 'org:bar:1.0' but it is not part of the resolved modules"))
+        failure.assertHasCause("Dependency lock for configuration 'lockedConf' is out of date:\n" +
+            "\tLocking constraint was not resolved: 'org:bar:1.0'")
```
Comment (melix) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy]: :o: The message is IMO too focuses on an implementation detail ("locking constraint") and should rather explain from a user point of view what is happening:

> Dependency graph doesn't contain 'org:bar:1.0' which used to be found in the lock file

or something similar.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java [hunk 3]
```diff
@@ -67,19 +69,27 @@ public BuildableLocalConfigurationMetadata addConfiguration(String name, String
...
 
         @Override
         void maybeAddGeneratedDependencies(ImmutableList.Builder<LocalOriginDependencyMetadata> result) {
-            if (canBeLocked) {
-                for (DependencyConstraint dependencyConstraint : dependencyLockingProvider.findLockedDependencies(getName())) {
-                    ModuleComponentSelector selector = DefaultModuleComponentSelector.newSelector(
-                        dependencyConstraint.getGroup(), dependencyConstraint.getName(), dependencyConstraint.getVersionConstraint());
-                    result.add(new LocalComponentDependencyMetadata(getComponentId(), selector, getName(), getAttributes(), null,
-                        Collections.<IvyArtifactName>emptyList(), Collections.<ExcludeMetadata>emptyList(), false, false, true, true, dependencyConstraint.getReason()));
+            if (locked) {
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java]: :o: I am confused as to why we need both `locked` and `isLockDefined`.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/locking/DependencyLockingArtifactVisitor.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import org.gradle.internal.component.local.model.LocalFileDependencyMetadata;
+import org.gradle.internal.component.local.model.RootConfigurationMetadata;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+public class DependencyLockingArtifactVisitor implements DependencyArtifactsVisitor {
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/locking/DependencyLockingArtifactVisitor.java]: :x: Would be nice to have some unit test coverage for this guy.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/locking/EmptyLockConstraint.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Override
+    public boolean isLockDefined() {
+        return false;
+    }
+
+    @Override
+    public Set<DependencyConstraint> getLockedDependencies() {
+        throw new UnsupportedOperationException("Cannot request locked dependencies when none are defined");
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/locking/EmptyLockConstraint.java]: :o: I feel like we can avoid this `EmptyLockConstraint` altogether. Could be a default lock constraint with an empty set of constraints, or something else, or maybe we don't need it. But it feels like accidental noise.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/locking/LockOutOfDateException.java [hunk 6]
```diff
@@ -16,28 +16,16 @@
...
-import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
-
-import java.util.Collection;
-
 public class LockOutOfDateException extends RuntimeException {
 
-    public static LockOutOfDateException createLockOutOfDateException(Iterable<String> errors) {
-        StringBuilder builder = new StringBuilder("Dependency lock out of date:\n");
+    public static LockOutOfDateException createLockOutOfDateException(String configurationName, Iterable<String> errors) {
+        StringBuilder builder = new StringBuilder("Dependency lock for configuration '" + configurationName + "' is out of date:\n");
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/locking/LockOutOfDateException.java]: You might want to use `TreeFormatter` here.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/LockConstraint.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+package org.gradle.api.internal.artifacts.dsl.dependencies;
+
+import org.gradle.api.artifacts.DependencyConstraint;
+
+import java.util.Set;
+
+/**
+ * Described the lock state for a given configuration.
+ */
+public interface LockConstraint {
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/LockConstraint.java]: It feels more like a `DependencyLockingState` than a `LockConstraint`.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/LockConstraint.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Indicates if there is locking in place
+     *
+     * @return {@code true} if locking is in place
+     */
+    boolean isLockDefined();
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/LockConstraint.java]: This really indicates if a lock file was found, as this is code exercised only if locking was enabled.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy (hunk 1 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy]: Changing

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy (hunk 2 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/locking/DependencyLockingIntegrationTest.groovy]: Changing

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java (hunk 3 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java]: It is really about the different state of the system:
* Locking enabled - yes / no
* Lock file exists - yes / no
* Lock file contains modules - yes / no

And I need all these differences at different places.
But it is true that here, I could be more lax as I only care about applying the constraints iff locking is enabled. I do not care much about the exist vs. empty state.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/locking/DependencyLockingArtifactVisitor.java (hunk 4 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/locking/DependencyLockingArtifactVisitor.java]: Adding

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java (hunk 3 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/RootLocalComponentMetadata.java]: I see, I would rename the variables/methods for clarity, then.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/locking/EmptyLockConstraint.java (hunk 5 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/locking/EmptyLockConstraint.java]: Removed

**Comment by ljacomet** [comment]
@melix I have addressed all comments I believe and did some more cleanup.

**Pull request merged**